src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RecepProducerService.java

```diff
@@ -1,20 +1,140 @@
+        var recepMessage = new RecepMessage(edifact);
+        return new Recep(recepMessage);
+    }
+
+    private String toEdifact(List<Segment> segments) {
+        return segments.stream()
+            .map(Segment::toEdifact)
+            .collect(Collectors.joining("\n"));
+    }
+
+    private List<Segment> mapEdifactToRecep(Interchange receivedInterchangeFromHa) throws EdifactValidationException {
+        List<Segment> recepMessageSegments = new ArrayList<>();
+
+        var recepInterchangeHeader = mapToRecipInterchangeHeader(receivedInterchangeFromHa);
+        var recepMessageHeader = mapToRecipMessageHeader(receivedInterchangeFromHa);
+        var recepMessageTrailer = mapToRecepMessageTrailer(2);//TODO: change when multiple messages handling is implemented
+        var recepInterchangeTrailer = mapToRecepInterchangeTrailer(receivedInterchangeFromHa);
+
+        recepMessageSegments.add(recepInterchangeHeader);
+        recepMessageSegments.add(recepMessageHeader);
+        recepMessageSegments.add(mapToRecepBeginningOfMessage(receivedInterchangeFromHa));
+        recepMessageSegments.add(mapToRecepNameAndAddress(receivedInterchangeFromHa));
+        recepMessageSegments.add(mapToRecepTranslationDateTime(receivedInterchangeFromHa));
+        recepMessageSegments.add(mapToReferenceMessageRecep(receivedInterchangeFromHa));
+        recepMessageSegments.add(mapToReferenceInterchangeRecep(receivedInterchangeFromHa));
+        recepMessageSegments.add(recepMessageTrailer);
+        recepMessageSegments.add(recepInterchangeTrailer);
+
+        recepMessageSegments.forEach(Segment::preValidate);
+
+        var recepInterchangeSequence = sequenceService.generateInterchangeId(
+            recepInterchangeHeader.getSender(),
+            recepInterchangeHeader.getRecipient());
+        var recepMessageSequence = sequenceService.generateMessageId(
+            recepInterchangeHeader.getSender(),
+            recepInterchangeHeader.getRecipient());
+
+        recepInterchangeHeader.setSequenceNumber(recepInterchangeSequence);
+        recepInterchangeTrailer
+            .setSequenceNumber(recepInterchangeSequence)
+            .setNumberOfMessages(1); //TODO: change when multiple messages handling is implemented
+
+        recepMessageHeader.setSequenceNumber(recepMessageSequence);
+        recepMessageTrailer
+            .setSequenceNumber(recepMessageSequence)
+            .setNumberOfSegments(7); //TODO: change when multiple messages handling is implemented. How many segments are there in message in total, including header and trailer
+
+        recepMessageSegments.forEach(Segment::validate);
+
+        return recepMessageSegments;
+    }
+
+    private InterchangeTrailer mapToRecepInterchangeTrailer(Interchange receivedInterchangeFromHa) {
+        return new InterchangeTrailer(receivedInterchangeFromHa.getInterchangeTrailer().getNumberOfMessages());
+    }
+
+    private MessageTrailer mapToRecepMessageTrailer(int refSegmentCount) {
```

Comment:
is this method needed? refSegmentCount is always 2

Reply:
currently yes
it will be variable when we support multiple messages